### ISSUE TITLE
What goal do you have?

### ISSUE THREAD

**Issue by charlenni** [jefffhaynes/LocalMap#1]
The title say it. What do you want to make? A rasterizer in C#? Or a rasterizer for a special map library? A vector tile source for a real vector map?

Btw. the first tile looks ok for the first shot. Try to change the rendering process as suggested.

**Comment by jefffhaynes** [comment]
I need a rasterizer that will work specifically in UWP.  At some point I may tackle full vector support but I was planning on starting with tile rastering.  I'm certainly open if you have any suggestions.

I have changed the approach to what you suggested.  I believe the problem now is that I'm not actually respecting "line-gap-width".  I'm currently working on the best approach for doing that in Win2D but I'll commit what I have with your changes.

**Comment by charlenni** [comment]
Further images are welcome 🙂

**Comment by jefffhaynes** [comment]
![image](https://user-images.githubusercontent.com/9734727/59161641-5ddc2b80-8ab3-11e9-8784-37b2da64ff05.png)

As you can see, at certain zoom levels I get what appear to be very thick (in this case, white) lines.  I'm currently trying to track down the source of that...

Are you able to get this running on your machine?

**Comment by charlenni** [comment]
I assume, that the selected style isn‘t very good for this use case. Perhaps try the [OSM Liberty](https://github.com/maputnik/osm-liberty) style. It has better colors. Than post the result.

No, I didn‘t tried your version. Up to now I only read the code.

**Comment by jefffhaynes** [comment]
![image](https://user-images.githubusercontent.com/9734727/59161813-3b4b1200-8ab5-11e9-8829-f0f2621d27d3.png)

**Comment by charlenni** [comment]
Good, looks better.

I assume, that you have problems with your pen width for outlines (the gray parts around the buildings) and the fill color of your polygons. Check both.

**Comment by charlenni** [comment]
And there are black lines inside the yellow roads. This aren‘t correct. Perhaps a debug option of your code?

**Comment by jefffhaynes** [comment]
I agree.  That's what I'm looking at now but so far I can't find the issue.  It may be my interpolation...

**Comment by charlenni** [comment]
Perhaps at https://github.com/jefffhaynes/LocalMap/blob/21989cef79f5704978ce29f14891bce177ac32ab/LocalMap/TileRasterer.cs#L217. Is the line size correct?

**Comment by jefffhaynes** [comment]
![image](https://user-images.githubusercontent.com/9734727/59162038-62571300-8ab8-11e9-85c2-d4fc2755f80b.png)

Oh geez, good catch

**Comment by charlenni** [comment]
I’m not sure, if you respect the line endings correct. Look in the upper left corner. The yellow road has a small gap at 7 o’clock.

**Comment by jefffhaynes** [comment]
I'm not sure.  I think I'm following style there but I see what you mean.  There seems to be an issue with closed polygons and how I'm drawing them though.  I don't know if that's related but I'm trying to track it down.

![image](https://user-images.githubusercontent.com/9734727/59162232-73555380-8abb-11e9-82fb-d2d1d0a7b32c.png)

**Comment by jefffhaynes** [comment]
Interestingly, there kind of seems to be an issue with the map here.  Here's mine:

![image](https://user-images.githubusercontent.com/9734727/59162415-ef04cf80-8abe-11e9-8550-4a4728bb3c92.png)

and here's Maputnik:

![image](https://user-images.githubusercontent.com/9734727/59162425-02179f80-8abf-11e9-84bd-1746e836beb0.png)

You can see that Maputnik isn't fully drawing the polygon section below the top

**Comment by jefffhaynes** [comment]
Fixed the errant line issue...

![image](https://user-images.githubusercontent.com/9734727/59162710-f67aa780-8ac3-11e9-81a1-ab8e3fa0fb2c.png)

**Comment by charlenni** [comment]
I assume, that at line https://github.com/jefffhaynes/LocalMap/blob/d8f1dd8d8ddbbaf4b7481d1a4c9f2192b0aa11f5/MapboxStyle/ColorFunctionConverter.cs#L25 the color conversion isn’t correct. The color is given as a hsla() function. Because it isn’t implemented, it returns black/null.

**Comment by jefffhaynes** [comment]
I'm not sure what you mean.  Hsla should be implemented...

https://github.com/jefffhaynes/LocalMap/blob/d8f1dd8d8ddbbaf4b7481d1a4c9f2192b0aa11f5/MapboxStyle/ConverterExtensions.cs#L128

**Comment by jefffhaynes** [comment]
My problem is that now that I fixed the line issue most roads are missing...

**Comment by charlenni** [comment]
Could you check, if hsla() returns correct values.

**Comment by jefffhaynes** [comment]
Yes, it seems to

**Comment by jefffhaynes** [comment]
I think my check to see if style is valid for a feature is wrong.

Yes, I didn't do what you originally said correctly.  Let me fix it...

**Comment by jefffhaynes** [comment]
Nope, I fixed it but still no roads...ok, taking a break.  Thanks for your help

**Comment by jefffhaynes** [comment]
Can you tell me, do I filter feature layers on source-layer or id or something else?

**Comment by charlenni** [comment]
I assume, that all parts, that use a hsla() color are drawn in black. If I check the map you published with Mapnik, than all this parts have a hsla() or hsl() color.

**Comment by jefffhaynes** [comment]
Ok maybe you’re right. I’ll check again


________________________________

**Comment by jefffhaynes** [comment]
Got it!  I had a few bugs related to filtering.  I'm not sure if there was a bug with hsla processing but I did rewrite it and clean it up.  Check it out now...

![image](https://user-images.githubusercontent.com/9734727/59224177-c4d50f80-8b9b-11e9-8fc7-1a9aeb0a3f63.png)

**Comment by charlenni** [comment]
Looks better, but the outline color of buildings isn’t correct. The code in the style file is

````
    {
      "id": "building",
      "type": "fill",
      "metadata": {},
      "source": "openmaptiles",
      "source-layer": "building",
      "minzoom": 13,
      "maxzoom": 14,
      "paint": {
        "fill-color": "hsl(35, 8%, 85%)",
        "fill-outline-color": {
          "base": 1,
          "stops": [
            [
              13,
              "hsla(35, 6%, 79%, 0.32)"
            ],
            [
              14,
              "hsl(35, 6%, 79%)"
            ]
          ]
        }
      },
      "layout": {
        "visibility": "visible"
      }
    }
````

I assume, that you perhaps have a problem with the stops?

**Comment by jefffhaynes** [comment]
It was another drawing bug

![image](https://user-images.githubusercontent.com/9734727/59229170-b42a9680-8ba7-11e9-8fd2-921bb854caf9.png)

**Comment by jefffhaynes** [comment]
I still have bugs somewhere in my path labeling (that's super tricky) but here it is with some fixes

![image](https://user-images.githubusercontent.com/9734727/59231275-1ab2b300-8bae-11e9-96f1-322ba42d5957.png)

**Comment by jefffhaynes** [comment]
![image](https://user-images.githubusercontent.com/9734727/59236591-f4e4d880-8bc4-11e9-9e59-fd9de0db07d8.png)

Path labeling!

**Comment by jefffhaynes** [comment]
Fixed that pattern issue...

![image](https://user-images.githubusercontent.com/9734727/59238306-822b2b80-8bcb-11e9-959e-9d23f1ed3b3f.png)

**Comment by xivk** [comment]
Kudos on this work, this is difficult and the results look awesome. I did this once in OsmSharp but it never worked as I wanted it to. :+1:

**Comment by jefffhaynes** [comment]
Thanks!  I’m still working on a new layout layer optimized for UWP but I’m actually struggling more with that. I think I’m on approach #11 now...

The tile drawing is still missing symbols and probably a few other things but it’s getting there.


________________________________

**Comment by charlenni** [comment]
You could find a ready approach for a tile server ( that‘s what you make) [here](https://github.com/AliFlux/VectorTileRenderer). It is a good base too. If you want a tile source for Mapsui, which convert tiles into features, than there is my [first try](https://github.com/charlenni/Mapsui.VectorTiles).

All three above lake of a big part of a map: labeling. You have to decide, which labels are draw and which not. There must be an order. Additionally labels, that are on another tile could overlap to this tile. And after all this efforts, you only get a bitmap tile, no vector tile.

I‘m try to overcome the last disadvantage by creating tile as SkPicture, a drawable vector tile, that could be drawn in all zoom levels on a canvas. But I‘m could get a way for the labeling.

**Comment by jefffhaynes** [comment]
I think the drawing is going ok. I’m sort of working on a replacement for Mapsui for UWP only. I need something with a little more flexibility than what is currently possible with Mapsui.


________________________________

**Comment by charlenni** [comment]
In which way?

**Comment by jefffhaynes** [comment]
I think Mapsui is obviously very flexible and portable so I don't mean to detract from that.  However, there are a few areas I would like to focus on for my project.

- I don't need portability so I believe I can optimize for UWP over some of the assumptions that Mapsui is forced to make.
- While it is possible to add widgets, pois, etc. to Mapsui the interface is somewhat complicated.
- Lastly (and this is just me being pedantic), the Mapsui API doesn't conform to many of the C# conventions for API design and naming.

But again, please don't take any of this the wrong way.  It is clearly an impressive project and there is a good chance I will end up using it when my own efforts fail :)

**Comment by charlenni** [comment]
No problem. I asked to learn, what could be made better.

The first two points are a problem of flexibility. So I assume, that it is impossible to change something in this part. But the third seems to worth a look. Do you have examples, where Mapsui isn’t correct?
